### ISSUE TITLE
Incorrectly inferred type for union types

### ISSUE THREAD

**Issue by RomanRolginSky** [microsoft/TypeScript#43962]
# Bug Report

### 🔎 Search Terms

infer, never, union types, error TS2345

### 🕗 Version & Regression Information

- This changed between versions 3.5.1 and all newer versions up to 4.3.0-beta. 3.5.1 doesn't throw such error, there is the error in 3.6.2
  
### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?jsx=0&ts=3.6.3#code/C4TwDgpgBAKgFgSwHYHMA8MB8UC8UDeAUFCVAG4CGANgK4QBcsA3MaWApGgIIA0UAQnwDCmABStSpAPZgujAPJgMfLph4TJUGfwVLeAtRskyhutIKgiNASkbxk6ESwC+LUJCiK0AST7zseKLAiKh2IejemNa42PaoaP4shAAmEADGVBQATtAAZjRIacAIUkhQaTkUwBBx6FiiOQDONFTAdraw4RiYLKkZ2XkFRSVlALYUSjB8AEpiYFlSAFbpbVCilLQMsNE42NMdXlNQs73pmTlQ+YXFpVDAE92iSBAAHqvr1HTtMeRSCMkHSZ8LBJZDVLK5ChpaAAdSyE0gyTQADVsERJMkqhRGMiWJJ5jJGgB+RhSABGyyKLhSZwGlyGNzKCEacIRECR9XJi0YFCQIA6XKgzKgrLAiO6p36Fyuw1uBQA7vDJmIuXYoAAfEVK8VYDowJLuWHa9neJC5CBZHLJAAiWO6uFgUFe1SQyUaWrZSOQ5qyx2wROOUDsbnARs9prANGAKICax2sS6qKSfXOg2uIygFGSyQAyhBgMVUI0Y+IMVicRqPWL2TH1PiFmBGqSKStCB1ReKOyazRarbb7ijK12kajMD1CIQ0qVGsB6enSjCEMEYOAHN5Gs1oHgS5CilIsiBGMOI1HB5rhzGomEHDGfujSDlgDQsmVd8B9yBRNYAHTsSCicYwFEUQlwgUZ4ygBUlRA6pwOsPhAJgsCflA8CEImJDRh+KCMNQ6x8OpSdp1nYJmXXGBshQfNkU+LcoAARiSKckBnTMABIHRlRlF2XVdUHXTdgIg+8SGY1jSNQB0KggKoanCIJEEacjKOo2jrDxB982fMoJJQX8OAgUtNBIRDRExe4IKzXN80LFBGjMrE+HwTM+QABQbRh6KgZx8LrYyoFM8yKAgoL4KMUh7iA9YILIawbBcdSJwoNjvw2OgmCgAB6TKoEaOApBaZIoDJaApFyO5QzuRTlKyKjgBozYgA)

### 💻 Code

```ts
type Thing<T> = {
    value: T;
    pipe<A, B, C>(
        opA: Op<T, A>,
        opB: Op<A, B>,
        opC: Op<B, C>
    ): Thing<C>;
};
type Op<I, O> = (thing: Thing<I>) => Thing<O>;

declare function createThing<T>(result: T): Thing<T>;
declare function map<T, R>(project: (value: T) => R): Op<T, R>;
declare function tap<T>(next: (value: T) => void): Op<T, T>;

interface Wrapped<V> {
    data: V;
    props?: object;
}
declare function isWrapped<T>(obj: any): obj is Wrapped<T>;
declare function unwrap<T>(obj: T | Wrapped<T>): T;

type WrappedInferredData<T> = T extends Wrapped<infer R> ? R : T;
type WrappedInput<V> = () => Thing<V>;

declare function addSettings<V>(
    data: V | Wrapped<V>,
    props: object
): Wrapped<WrappedInferredData<V | Wrapped<V>>>;

const functionWithTypingIssue = <V>(factory: WrappedInput<V | Wrapped<V>>): Thing<V> => {
    return factory().pipe(map((item) => unwrap(item)), map(item => item), map(item => unwrap(item)));
}

const thisIsTargetValue = 1;

const a$ = functionWithTypingIssue(() => {
    const thing = createThing(thisIsTargetValue);
    return thing.pipe(
        map((data) => addSettings(data, { anyProp: 1 })),
        map((data) => data), // error here
        tap((v) => v)
    );
});

a$.value; // should be of type thisIsTargetValue
```
```
Argument of type 'Op<Wrapped<number>, Wrapped<number>>' is not assignable to parameter of type 'Op<Wrapped<number>, Wrapped<never>>'.
  Type 'Wrapped<number>' is not assignable to type 'Wrapped<never>'.
    Type 'number' is not assignable to type 'never'. 
```

### 🙁 Actual behavior

Type incorrectly infers from union type 

### 🙂 Expected behavior

Type should be resolved correctly. `typeof thisIsTargetValue === number` for this example

**Comment by RyanCavanaugh** [comment]
I'm not sure if this is a bug or not; a much shorter repro would be appreciated.

**Comment by RomanRolginSky** [comment]
@RyanCavanaugh I removed some non-required declaration but even shorter is unreal 😊  Part of declarations is a mock of rxjs-library (like `Op`, `Thing`, `map`,`tap`) so it can't be simplified/removed

```ts
type Op<I, O> = (thing: Thing<I>) => Thing<O>;
type Thing<T> = {
    value: T;
    pipe<A, B, C>(
        opA: Op<T, A>,
        opB: Op<A, B>,
    ): Thing<B>;
};
type Wrapped<V> = { data: V }
type WrappedInferredData<T> = T extends Wrapped<infer R> ? R : T;

declare function createThing<T>(result: T): Thing<T>;
declare function map<T, R>(project: (value: T) => R): Op<T, R>;
declare function tap<T>(next: (value: T) => void): Op<T, T>;
declare function addSettings<V>(data: V | Wrapped<V>): Wrapped<WrappedInferredData<V | Wrapped<V>>>;
declare function functionWithTypingIssue<V>(factory: () => Thing<V | Wrapped<V>>): Thing<V>;

const thisIsTargetValue = 1;

const result = functionWithTypingIssue(() => {
    const thing = createThing(thisIsTargetValue);
    return thing.pipe(
        map((data) => addSettings(data)),
        tap(v => v)
    );
});

result.value; // should be of type thisIsTargetValue
```
[Typescript playground](https://www.typescriptlang.org/play?jsx=0&ts=4.1.5#code/C4TwDgpgBA8mA8BJANLAfFAvFAFMAFgJYB2A5gFxQAqRZSaAlFhjSafDGgNwBQok1WuyoZsAbx5QpUAG4BDADYBXCJSq9pUMIUjwAgqgBCqAMJockzdID2YPZTjwqqPWmSWrUW4YcIDUQzcPKAY1IXhA3gBfXn5oAHUAJzkwSAATeAA1USgxKDS5YDlKTKgovnAE5NSINMRiADMIRMTagBFCuSccqigIAA9gCGI0gGcoJJT0+BImxKgAJQwAfkWoNV4eNIgAYwU5VqgGpWId4EJrYigd1sKIVjoRHFbRpQVgNVDBNm7ebb2DtBjqdzpcoABbFJOVBLHBgRLWABWuw+uHkylU1CYmAwCy+jmci24W12+0OwLOFyuRQQT2IA1ROHRKk+zFk1kIaXxtNQIj+pMBRxOlLBcjSaQAyhBgOcyKMsuYCkUSlAAD4TarTbJfSY1DK69L1OatNIdIpZNUaqa1BVoYn-MlA4Wgq4Ul3xQgEKjgNiIUavCAKnANORnayJECUHDYljhUrqg027KMMI-bKbHaXUbAKAEQijP1UA6kaWZRQqLBQACMGazOZebxz2DdVI9Xp9ZD9AZw0bZEk0meI2dzQkrNwgdwepDwRALoyLiRLwDLGIYGmkrWASkS1KEADptJALJ4pJCwD2lXIY1AxZLpbLSKMcJeGAx3CfcykmWyZAxgmueCiACeAbd492ZCAuCgAB6aCoFGfBrDeNIoAAI2gawGlzSoR3zQti1LcsICAA)

**Comment by jakebailey** [comment]
Thanks for providing the versions, I was able to bisect this to #31662.

**Comment by jakebailey** [comment]
Just to give the test case I'm actually testing here: [Playground Link](https://www.typescriptlang.org/play?#code/C4TwDgpgBA8mA8BJANLAfFAvFAFMAFgJYB2A5gFxQAqRZSaAlFhjSafDGgNwBQok1WuyoZsAbx5QpUAG4BDADYBXCJSq9pUMIUjwAgqgBCaHJM3SA9mD2U48Kqj1pkZ81CuHbCA1GMvNDGpC8Ma8AL68-NAA6gBOcmCQACbwAGqiUGJQSXLAcpSpUBE8PEkQAMYKcrHQ5RbEAM7AUARsQWzwxEoAtgBGELHcJWWV1dAAZkrE5cCE9VDdCfaoAEomYLEWAFYVwJQ48sqq1EyYGCuBsAgOUGu8I1U1UJPTs-N51ybEEAAee7iHFRqU4YGQWQhJS52G4ie4VR4TKYzObEKAAd3iCHSOByeQKlziCWSaSGDzGzyRb1R5RquQgemISQAqsQMUtseM5DMLLEQPsQYIOoUAD5QQmJCApdKMdp0dJw0ZPF7I+YKCykA6KIFQOTEECXMEQ3g8OqNZo1BpKBTAACMWCgNIgdIZzNZmJwOAFrTIADptJBTJpFmAPbi5AK2SGwwwGP5pB8PTIBWqNUnYzwY8bTU0oBarcAAEz2x3OxksyMer1CP06CCB+MJRPJ9UHGNxqTB0O5cPMdHu6PpzM8IA)

(It's somewhat smaller but not much)
